mri/reconstruct/fourier.py

```diff
@@ -207,5 +231,284 @@ def adj_op(self, x):
+                    np.copy(img[ch]).astype(dtype))
+                gx = self.nufftObj.thr.copy_array(self.nufftObj.x_Nd)
+                # Forward operator of the NUFFT
+                gy = self.nufftObj.forward(gx)
+                y.append(np.squeeze(gy.get()))
+            y = np.asarray(y)
+        return y * 1.0 / np.sqrt(np.prod(self.Kd))
+
+    def adj_op(self, x):
+        """ This method calculates inverse masked non-uniform Fourier
+        transform of a 1-D coefficients array.
+        Parameters
+        ----------
+        x: np.ndarray
+            masked non-uniform Fourier transform 1D data.
+        Returns
+        -------
+        img: np.ndarray
+            inverse 3D discrete Fourier transform of the input coefficients.
+        """
+        if self.nb_coils == 1:
+            dtype = np.complex64
+            cuda_array = self.nufftObj.thr.to_device(x.astype(dtype))
+            gx = self.nufftObj.adjoint(cuda_array)
+            img = np.squeeze(gx.get())
+        else:
+            dtype = np.complex64
+            img = []
+            for ch in range(self.nb_coils):
+                cuda_array = self.nufftObj.thr.to_device(np.copy(
+                    x[ch]).astype(dtype))
+                gx = self.nufftObj.adjoint(cuda_array)
+                img.append(gx.get())
+            img = np.asarray(np.squeeze(img))
+        return img * np.sqrt(np.prod(self.Kd))
+
+
+class NonCartesianFFT(FourierBase):
+    """This class wraps around different implementation algorithms for NFFT"""
+    def __init__(self, samples, shape, implementation='cpu', n_coils=1):
+        """ Initialize the class.
+        Parameters
+        ----------
+        samples: np.ndarray (Mxd)
+            the samples locations in the Fourier domain where M is the number
+            of samples and d is the dimensionnality of the output data
+            (2D for an image, 3D for a volume).
+        shape: tuple of int
+            shape of the image (not necessarly a square matrix).
+        implementation: str 'cpu' | 'cuda' | 'opencl', default 'cpu'
+            which implementation of NFFT to use.
+        n_coils: int default 1
+            Number of coils used to acquire the signal in case of multiarray
+            receiver coils acquisition
+        """
+        self.shape = shape
+        self.samples = samples
+        self.nb_coils = n_coils
+        if implementation == 'cpu':
+            self.implementation = NFFT(samples=samples, shape=shape,
+                                       n_coils=self.nb_coils)
+        elif implementation == 'cuda' or implementation == 'opencl':
+            self.implementation = NUFFT(samples=samples, shape=shape,
+                                        platform=implementation,
+                                        n_coils=self.nb_coils)
+        else:
+            raise ValueError('Bad implementation ' + implementation +
+                             ' chosen')
```

Comment:
I'm not sure that this is a good message error, I would recommend to add the different options inside message

Reply:
Yes, will add options